app/controllers/supplementary/PreviousDocumentsController.scala

```diff
@@ -32,30 +34,64 @@ import views.html.supplementary.previous_documents
 import scala.concurrent.{ExecutionContext, Future}
 
 class PreviousDocumentsController @Inject()(
-  appConfig: AppConfig,
   override val messagesApi: MessagesApi,
   authenticate: AuthAction,
+  errorHandler: ErrorHandler,
   customsCacheService: CustomsCacheService
-)(implicit ec: ExecutionContext)
+)(implicit ec: ExecutionContext, appConfig: AppConfig)
     extends FrontendController with I18nSupport {
 
   def displayForm(): Action[AnyContent] = authenticate.async { implicit request =>
-    customsCacheService.fetchAndGetEntry[Document](supplementaryCacheId, formId).map {
-      case Some(data) => Ok(previous_documents(appConfig, form.fill(data)))
-      case _          => Ok(previous_documents(appConfig, form))
+    customsCacheService.fetchAndGetEntry[PreviousDocumentsData](supplementaryCacheId, formId).map {
+      case Some(data) => Ok(previous_documents(form, data.documents))
+      case _          => Ok(previous_documents(form, Seq.empty))
     }
   }
 
   def savePreviousDocuments(): Action[AnyContent] = authenticate.async { implicit request =>
-    form
-      .bindFromRequest()
-      .fold(
-        (formWithErrors: Form[Document]) =>
-          Future.successful(BadRequest(previous_documents(appConfig, formWithErrors))),
-        validForm =>
-          customsCacheService.cache[Document](supplementaryCacheId, formId, validForm).map { _ =>
-            Redirect(controllers.supplementary.routes.GoodsItemNumberController.displayForm())
+    import MultipleItemsHelper._
+
+    val boundForm = form.bindFromRequest()
+
+    val actionTypeOpt = request.body.asFormUrlEncoded.map(FormAction.fromUrlEncoded(_))
+
+    val cachedData = customsCacheService
+      .fetchAndGetEntry[PreviousDocumentsData](supplementaryCacheId, formId)
+      .map(_.getOrElse(PreviousDocumentsData(Seq.empty)))
+
+    cachedData.flatMap { cache =>
```

Comment:
In the name of `encapsulation` and readability, I would extract the `Add`, `Remove` and `Continue` implementation to it's own methods.
Not a blocker ;)

Reply:
I still :heart: this approach! :)